lib/base.rb

```diff
@@ -0,0 +1,28 @@
+module Ipfs
+  class Base16
+    ALPHABET = '0123456789abcdef'
+
+    def self.decode(hexdigits)
+      valid?(hexdigits) \
+        ? convert_to_byte_array(hexdigits) \
+        : []
+    end
+
+    def self.convert_to_byte_array(hexdigits)
+      convert_to_byte = lambda do |hexdigits|
+        hexdigits
+          .each_char
+          .reduce(0) { |byte, hexdigit| byte * 16 + ALPHABET.index(hexdigit) }
+      end
+
+      hexdigits
+        .scan(/.{1,2}/)
+        .map { |hex_digits_pair| convert_to_byte.(hex_digits_pair) }
+    end
+
+    def self.valid?(hexdigits)
+      regex = /\A[\d#{ALPHABET}]+\z/i
+      hexdigits.match?(regex)
```

Comment:
The regex shouldn't be directly put in the method call instead of being saved in the `regex` variable?

Reply:
I wanted to do that at first, but i thought it would make the method call look less ugly

Reply:
indeed, but on it adds nothing to the comprehension of the call.

In the perf. point of view it requires a registration in memory that is absolutely not needed knowing that the `regex` variable is not reused after the call.